File: archive-triage-runner/python/src/archive_triage/triage_service.py
"""Orchestrates the per-file pipeline and streams :class:`RowResult` values.

Knows nothing about HTTP or the filesystem — delegates to
:class:`EvageneApi` for API calls and receives pre-read ``(path, content)``
pairs from the scanner.  Errors are captured as rows rather than raised
so a single rotten file does not sink the whole archive.
"""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from typing import Any

from .evagene_client import EvageneApi, EvageneApiError
from .gedcom_scanner import GedcomFile
from .row_result import RowResult


@dataclass(frozen=True)
class TriageOptions:
    concurrency: int


class TriageService:
    def __init__(self, client: EvageneApi, options: TriageOptions) -> None:
        self._client = client
        self._options = options

    def triage(self, files: Iterable[GedcomFile]) -> Iterator[RowResult]:
        with ThreadPoolExecutor(max_workers=self._options.concurrency) as pool:
            yield from pool.map(self._triage_one, files)

    def _triage_one(self, file: GedcomFile) -> RowResult:
        display_name = file.path.stem
        try:
            pedigree_id = self._client.create_pedigree(display_name)
        except EvageneApiError as error:
            return _failure(
                pedigree_id="",
                proband_name=display_name,
                message=f"create_pedigree failed: {error}",
            )

        try:
            self._client.import_gedcom(pedigree_id, file.content)
        except EvageneApiError as error:
            return _failure(pedigree_id, display_name, f"import_gedcom failed: {error}")

        try:
            if not self._client.has_proband(pedigree_id):
                return _failure(
                    pedigree_id,
                    display_name,
                    "no proband designated in GEDCOM — mark one with a _PROBAND 1 tag.",
                )
        except EvageneApiError as error:
            return _failure(pedigree_id, display_name, f"proband check failed: {error}")

        try:
            payload = self._client.calculate_nice(pedigree_id)
        except EvageneApiError as error:
            return _failure(pedigree_id, display_name, f"calculate_nice failed: {error}")

        return _row_from_payload(pedigree_id, display_name, payload)


def _row_from_payload(
    pedigree_id: str,
    fallback_name: str,
    payload: dict[str, Any],
) -> RowResult:
    cancer_risk = payload.get("cancer_risk")
    proband_name = _string(payload, "counselee_name") or fallback_name
    if not isinstance(cancer_risk, dict):
        return _failure(pedigree_id, proband_name, "NICE response missing cancer_risk block.")

    category = _string(cancer_risk, "nice_category")
    triggers = cancer_risk.get("nice_triggers", [])
    if not category or not isinstance(triggers, list):
        return _failure(pedigree_id, proband_name, "NICE response schema unexpected.")

    return RowResult(
        pedigree_id=pedigree_id,
        proband_name=proband_name,
        category=category,
        refer_for_genetics=_optional_bool(cancer_risk, "nice_refer_genetics"),
        triggers_matched_count=len(triggers),
        error="",
    )


def _failure(pedigree_id: str, proband_name: str, message: str) -> RowResult:
    return RowResult(
        pedigree_id=pedigree_id,
        proband_name=proband_name,
        category="",
        refer_for_genetics=None,
        triggers_matched_count=0,
        error=message,
    )


def _string(container: dict[str, Any], key: str) -> str:
    value = container.get(key)
    return value if isinstance(value, str) else ""


def _optional_bool(container: dict[str, Any], key: str) -> bool | None:
    value = container.get(key)
    return value if isinstance(value, bool) else None

File: archive-triage-runner/dotnet/src/ArchiveTriage/TriageService.cs
using System.Text.Json;

namespace ArchiveTriage;

public sealed record TriageOptions(int Concurrency);

public sealed class TriageService
{
    private readonly IEvageneApi client;
    private readonly TriageOptions options;

    public TriageService(IEvageneApi client, TriageOptions options)
    {
        this.client = client;
        this.options = options;
    }

    public async IAsyncEnumerable<RowResult> TriageAsync(
        IReadOnlyList<GedcomFile> files,
        [System.Runtime.CompilerServices.EnumeratorCancellation]
        CancellationToken cancellationToken = default)
    {
        using var throttle = new SemaphoreSlim(this.options.Concurrency);
        var tasks = files.Select(file => this.TriageOneAsync(file, throttle, cancellationToken)).ToList();

        foreach (var task in tasks)
        {
            yield return await task.ConfigureAwait(false);
        }
    }

    private async Task<RowResult> TriageOneAsync(
        GedcomFile file,
        SemaphoreSlim throttle,
        CancellationToken cancellationToken)
    {
        await throttle.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            return await this.PipelineAsync(file, cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            throttle.Release();
        }
    }

    private async Task<RowResult> PipelineAsync(GedcomFile file, CancellationToken cancellationToken)
    {
        var displayName = Path.GetFileNameWithoutExtension(file.Path);

        string pedigreeId;
        try
        {
            pedigreeId = await this.client.CreatePedigreeAsync(displayName, cancellationToken).ConfigureAwait(false);
        }
        catch (EvageneApiException ex)
        {
            return Failure(string.Empty, displayName, $"create_pedigree failed: {ex.Message}");
        }

        try
        {
            await this.client.ImportGedcomAsync(pedigreeId, file.Content, cancellationToken).ConfigureAwait(false);
        }
        catch (EvageneApiException ex)
        {
            return Failure(pedigreeId, displayName, $"import_gedcom failed: {ex.Message}");
        }

        try
        {
            var hasProband = await this.client.HasProbandAsync(pedigreeId, cancellationToken).ConfigureAwait(false);
            if (!hasProband)
            {
                return Failure(
                    pedigreeId,
                    displayName,
                    "no proband designated in GEDCOM — mark one with a _PROBAND 1 tag.");
            }
        }
        catch (EvageneApiException ex)
        {
            return Failure(pedigreeId, displayName, $"proband check failed: {ex.Message}");
        }

        JsonElement payload;
        try
        {
            payload = await this.client.CalculateNiceAsync(pedigreeId, cancellationToken).ConfigureAwait(false);
        }
        catch (EvageneApiException ex)
        {
            return Failure(pedigreeId, displayName, $"calculate_nice failed: {ex.Message}");
        }

        return RowFromPayload(pedigreeId, displayName, payload);
    }

    private static RowResult RowFromPayload(string pedigreeId, string fallbackName, JsonElement payload)
    {
        var probandName = OptionalString(payload, "counselee_name");
        if (probandName.Length == 0)
        {
            probandName = fallbackName;
        }

        if (!payload.TryGetProperty("cancer_risk", out var cancerRisk)
            || cancerRisk.ValueKind != JsonValueKind.Object)
        {
            return Failure(pedigreeId, probandName, "NICE response missing cancer_risk block.");
        }

        var category = OptionalString(cancerRisk, "nice_category");
        if (category.Length == 0)
        {
            return Failure(pedigreeId, probandName, "NICE response schema unexpected.");
        }

        var triggers = cancerRisk.TryGetProperty("nice_triggers", out var triggersElement)
            && triggersElement.ValueKind == JsonValueKind.Array
                ? triggersElement.GetArrayLength()
                : 0;

        bool? referForGenetics = null;
        if (cancerRisk.TryGetProperty("nice_refer_genetics", out var refer)
            && (refer.ValueKind == JsonValueKind.True || refer.ValueKind == JsonValueKind.False))
        {
            referForGenetics = refer.GetBoolean();
        }

        return new RowResult(
            pedigreeId,
            probandName,
            category,
            referForGenetics,
            triggers,
            string.Empty);
    }

    private static RowResult Failure(string pedigreeId, string probandName, string message) =>
        new(pedigreeId, probandName, string.Empty, null, 0, message);

    private static string OptionalString(JsonElement container, string key) =>
        container.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.String
            ? (value.GetString() ?? string.Empty)
            : string.Empty;
}
